extensions/data-plane/data-plane-google-bigquery/src/main/java/org/eclipse/edc/connector/dataplane/gcp/bigquery/pipeline/BigQueryDataSourceFactory.java

```diff
@@ -0,0 +1,104 @@
+ *
+ */
+
+package org.eclipse.edc.connector.dataplane.gcp.bigquery.pipeline;
+
+import com.google.auth.oauth2.GoogleCredentials;
+import org.eclipse.edc.connector.dataplane.gcp.bigquery.params.BigQueryRequestParamsProvider;
+import org.eclipse.edc.connector.dataplane.spi.pipeline.DataSource;
+import org.eclipse.edc.connector.dataplane.spi.pipeline.DataSourceFactory;
+import org.eclipse.edc.gcp.bigquery.BigQueryConfiguration;
+import org.eclipse.edc.gcp.bigquery.service.BigQueryServiceSchema;
+import org.eclipse.edc.gcp.common.GcpException;
+import org.eclipse.edc.gcp.iam.IamService;
+import org.eclipse.edc.spi.monitor.Monitor;
+import org.eclipse.edc.spi.result.Result;
+import org.eclipse.edc.spi.types.TypeManager;
+import org.eclipse.edc.spi.types.domain.transfer.DataFlowStartMessage;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.concurrent.ExecutorService;
+
+import static org.eclipse.edc.gcp.bigquery.service.BigQueryServiceSchema.BIGQUERY_DATA;
+
+/**
+ * Instantiates the BigQuery data transfer source {@link BigQueryDataSource} for requests whose
+ * source data type is BigQueryData. This factory is registered by the BigQuery Data Plane Extension.
+ */
+public class BigQueryDataSourceFactory implements DataSourceFactory {
+    private final BigQueryConfiguration configuration;
+    private final BigQueryRequestParamsProvider requestParamsProvider;
+    private final Monitor monitor;
+    private final TypeManager typeManager;
+    private final ExecutorService executorService;
+    private IamService iamService;
+
+    public BigQueryDataSourceFactory(BigQueryConfiguration configuration, Monitor monitor,
+                                     BigQueryRequestParamsProvider requestParamsProvider,
+                                     TypeManager typeManager, ExecutorService executorService,
+                                     IamService iamService) {
+        this.configuration = configuration;
+        this.monitor = monitor;
+        this.requestParamsProvider = requestParamsProvider;
+        this.typeManager = typeManager;
+        this.executorService = executorService;
+        this.iamService = iamService;
+    }
+
+    @Override
+    public boolean canHandle(DataFlowStartMessage message) {
+        return BigQueryServiceSchema.BIGQUERY_DATA.equals(message.getSourceDataAddress().getType());
+    }
+
+    @Override
+    public @NotNull Result<Void> validateRequest(DataFlowStartMessage message) {
+        // canHandle has been already invoked to have this factory selected.
+        // BigQuerySinkDataAddressValidator has already checked message.getDestinationDataAddress().
+        // BigQuerySourceDataAddressValidator has already checked message.getSourceDataAddress().
+
+        return Result.success();
+    }
+
+    @Override
+    public String supportedType() {
+        return BIGQUERY_DATA;
+    }
+
+    @Override
+    public DataSource createSource(DataFlowStartMessage message) {
+        if (!canHandle(message)) {
```

Comment:
better to invoke the validator here, because we must be sure that source and sink addresses are valid

Reply:
Source address validator is invoked as soon as the asset is added to the provider catalog, sink validator is created before the sink factory validates the request. createSource method is invoked after this already took place

Reply:
at this point in the code, we don't know anything about Assets or catalogs. Also, this is the source factory, so the sink factory is of no concern here, because it should only validate ... well ... the sink address. 

The source factory must validate and verify the source address on its own.

Reply:
Done